duckduckgo/duckduckgo-privacy-extension
# Full HTTPS coverage

**Reviewer:** @jdorweiler 

## Description:
We are adding support for remote Smarter Encryption service to cover even more websites.

## Steps to test this PR:

⚠️ Note that FF and Chrome implementations are slightly different (Chrome has to handle requests synchronously), so be sure to test in both browsers ✌️ 

1. Test https websites by going directly to their https versions:
- https://duckduckgo.com
- https://www.cnn.com/
- https://wikipedia.org

Those shouldn't be affected or even mentioned in the background page console.

2. Test https websites by going to their http versions:
- http://duckduckgo.com
- http://www.cnn.com/
- http://wikipedia.org

Those should be upgraded by our bloom filters ("HTTPS: Bloom filter - host is upgradable" in the bg page console)

3. Test http only websites that are not in our bloom filter e.g.
- http://burymewithmymoney.com/
- http://tinytuba.com/
- http://xinhuanet.com/
- http://soso.com/

Those are not in our bloom filters and will require checking with the SE service ("HTTPS Service: Requesting information for reverso.net "). You may see website being upgraded to https for a split of a second and then downgraded to http once response from the service returns. Subsequent loads should happen immediately, w/o us trying to upgrade because response should be cached ("HTTPS: Service cache - host is not upgradable").

4. Reload extension (to clear the cache), block access to the service (127.0.0.1 duckduckgo.com) and visit websites from #3 

Those should be broken because they are not in our bloom filters and when service is unavailable we default to HTTPS.

5. Test sites from https://staticcdn.duckduckgo.com/https/https-whitelist.json - those should NOT be upgraded ("HTTPS: Safelist - host is not upgradable") - note that those hosts can upgrade to HTTPS via redirect/HSTS themselves sometimes

6. Test sites from https://ddg-sandbox.s3.amazonaws.com/negative-whitelist.json - those should be upgraded ("HTTPS: Safelist - host is upgradable")

7. Make sure that turning off site privacy protection works and host is not updated

## Automated tests:

###### Reviewer Checklist:

###### PR Author Checklist:
